src/node/historical_queries.h

```diff
@@ -0,0 +1,341 @@
+      }
+
+      // TODO: Check the signature is produced by the claimed node (needs
+      // retrieval of the node's cert)
+      const auto node_info = get_node_info(sig->node);
+      if (!node_info.has_value())
+      {
+        throw std::logic_error(fmt::format(
+          "Signature {} claims it was produced by node {}: This node is "
+          "unknown",
+          sig_idx,
+          sig->node));
+      }
+
+      auto verifier = tls::make_verifier(node_info->cert);
+      const auto verified = verifier->verify_hash(
+        real_root.h.data(),
+        real_root.h.size(),
+        sig->sig.data(),
+        sig->sig.size());
+      if (!verified)
+      {
+        throw std::logic_error(
+          fmt::format("Signature at {} is invalid", sig_idx));
+      }
+
+      // TODO: Find which of our untrusted indices are in this tree
+      auto it = requests.begin();
+      while (it != requests.end())
+      {
+        auto& request = it->second;
+
+        if (request.current_stage == RequestStage::Untrusted)
+        {
+          const auto& untrusted_idx = it->first;
+          const auto& untrusted_hash = request.entry_hash;
+          const auto& untrusted_store = request.store;
+
+          try
+          {
+            const auto receipt = tree.get_receipt(untrusted_idx);
+            LOG_INFO_FMT(
+              "From signature at {}, constructed a receipt for {}",
+              sig_idx,
+              untrusted_idx);
+
+            // TODO: Check that the receipt matches our untrusted entry
+            // const auto& entry_hash_in_receipt = receipt.leaf;
+            // if (untrusted_hash != entry_hash_in_receipt)
+            // {
+            //   throw std::logic_error(
+            //     fmt::format("Hash mismatch for {}", untrusted_idx));
+            // }
+
+            // Move stores from untrusted to trusted
+            // TODO: Temp solution, blindly trust everything for now
+            LOG_INFO_FMT("Now trusting {}", untrusted_idx);
+            request.current_stage = RequestStage::Trusted;
+            ++it;
+          }
+          catch (const std::exception& e)
+          {
+            LOG_INFO_FMT(
+              "Signature at {} does not cover {}: {}",
+              sig_idx,
+              untrusted_idx,
+              e.what());
+            // TODO: Can we expose "what indices do you cover" through the
+            // MerkleTree, rather than try-catch?
+            ++it;
+            // TODO: Should we abandon this untrusted entry now?
+          }
+        }
+        else
+        {
+          ++it;
+        }
+      }
+    }
+
+    void deserialise_ledger_entry(
+      consensus::Index idx, const LedgerEntry& entry)
+    {
+      StorePtr store = std::make_shared<kv::Store>();
+
+      store->set_encryptor(source_store.get_encryptor());
+
+      // TODO: Add a lazy clone option?
+      store->clone_schema(source_store);
+      store->set_strict_versions(false);
```

Comment:
Since this is set once at construction time and never reset can this flag be passed to the ctor?